src/test/rpc/ApiExperiment.h

```diff
@@ -0,0 +1,47 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2021 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLED_APIEXPERIMENT_H
+#define RIPPLED_APIEXPERIMENT_H
+
+#include <ripple/rpc/impl/RPCHelpers.h>
+
+namespace ripple {
+namespace test {
+
+class ApiExperiment
+{
+public:
+    ApiExperiment() : cachedMaxVersion_(RPC::ApiMaximumSupportedVersion)
+    {
+        RPC::ApiMaximumSupportedVersion = RPC::ApiExperimentalVersion;
+    }
+
+    ~ApiExperiment()
+    {
+        RPC::ApiMaximumSupportedVersion = cachedMaxVersion_;
+    }
+
+private:
+    unsigned int cachedMaxVersion_;
+};
+
+}  // namespace test
+}  // namespace ripple
```

Comment:
I really like having a class that uses RAII to set and recover the `ApiMaximumSupportedVersion`.  Good call.

There are a few things I'm inclined to tweak about this implementation.  Let's see what you think.

1. I think the source code may be easier to read if you can see the experimental version you're getting right there in the source code.  So I'd like to see the (only) constructor take an `unsigned int` as an argument.  

2. For me, `ApiExperiment` does not really describe what we're doing.  What we're doing is temporarily setting a new MaxApiVersion.  So maybe we should call the class `TempMaxApiVersion`.  That, plus the first point, puts the following in the source code:

```
    test::TempMaxApiVersion maxV(2);
```

3. I think we want to discourage folks getting tricky with nesting these things or passing them around.  So we want to be sure to turn off copy constructors, assignment operators, and the like.

4. Rather than cache the previous version, I think it's a little bit safer to have the destructor simply jam the value back to the default value.  That way we can be _really_ confident that the value is set back the way we want no matter if someone tries to do something tricky with these objects.

So, given all those thoughts, here's what I ended up with:
```
class TempMaxApiVersion
{
public:
    explicit TempMaxApiVersion(unsigned int experimentalVersion)
    {
        if (experimentalVersion < RPC::ApiMinimumSupportedVersion ||
            experimentalVersion > RPC::ApiExperimentalVersion)
            Throw<std::range_error>(
                "Specified experimental API version outside valid range");

        RPC::WARNING_setApiMaximumSupportedVersion(experimentalVersion);
    }

    // A TempMaxApiVersion is not copyable or assignable.
    TempMaxApiVersion(TempMaxApiVersion const&) = delete;
    TempMaxApiVersion&
    operator=(TempMaxApiVersion const&) = delete;

    ~TempMaxApiVersion()
    {
        RPC::WARNING_setApiMaximumSupportedVersion(
            RPC::ApiDefaultMaximumSupportedVersion);
    }
};

#ifndef __INTELLISENSE__
static_assert(std::is_default_constructible_v<TempMaxApiVersion> == false);
static_assert(std::is_copy_constructible_v<TempMaxApiVersion> == false);
static_assert(std::is_move_constructible_v<TempMaxApiVersion> == false);
static_assert(std::is_copy_assignable_v<TempMaxApiVersion> == false);
static_assert(std::is_move_assignable_v<TempMaxApiVersion> == false);
static_assert(std::is_nothrow_destructible_v<TempMaxApiVersion> == true);
#endif
```
They are really just minor tweaks on your design.  But I thought you should at least see where I ended up.

I'm open to your thoughts.  Thanks.

Reply:
Thanks so much for the suggestions. They are great. 
Since we are gonna use a config option to promote the version number, this class is removed. But I will follow the suggestions in similar situations.